// @fitbit/fdb-protocol/lib/FDBTypes/Console.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var t = require("io-ts");
var Structures_1 = require("./Structures");
exports.ConsoleDebuggerCapabilities = t.partial({
    appLogging: t.boolean,
}, 'ConsoleDebuggerCapabilities');
exports.ConsoleMessageKind = t.union([
    t.literal('log'),
    t.literal('info'),
    t.literal('warn'),
    t.literal('error'),
]);
exports.ConsoleMessage = t.intersection([
    t.partial({
        timestamp: Structures_1.Timestamp,
        position: Structures_1.Position,
        fromHost: t.boolean,
    }),
    t.interface({
        emittedBy: Structures_1.InstalledAppComponent,
        kind: exports.ConsoleMessageKind,
        message: t.array(t.any),
    }),
], 'ConsoleMessage');
exports.TraceMessageKind = t.union([
    t.literal('trace'),
    t.literal('assert'),
    t.literal('exception'),
]);
exports.TraceMessage = t.intersection([
    t.interface({
        emittedBy: Structures_1.InstalledAppComponent,
        stack: t.array(Structures_1.Position),
        kind: exports.TraceMessageKind,
        message: t.array(t.any),
    }),
    t.partial({
        timestamp: Structures_1.Timestamp,
    }),
], 'TraceMessage');
