// @fitbit/fdb-protocol/lib/FDBTypes/Initialize.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var t = require("io-ts");
var AppInstall_1 = require("./AppInstall");
var BulkData_1 = require("./BulkData");
var Console_1 = require("./Console");
var Eval_1 = require("./Eval");
var HeapSnapshot_1 = require("./HeapSnapshot");
var Launch_1 = require("./Launch");
var Meta_1 = require("./Meta");
var Screenshot_1 = require("./Screenshot");
exports.DebuggerCapabilities = t.partial({
    protocol: Meta_1.ProtocolCapabilities,
    io: BulkData_1.IOCapabilities,
    console: Console_1.ConsoleDebuggerCapabilities,
    experimental: t.any,
}, 'DebuggerCapabilities');
exports.InitializeParams = t.intersection([
    t.interface({
        capabilities: exports.DebuggerCapabilities,
    }),
    t.partial({
        userAgent: t.string,
    }),
], 'InitializeParams');
exports.HostKind = t.union([
    t.literal('device'),
    t.literal('companion'),
]);
exports.AppDebugCapabilities = t.partial({
    heapSnapshot: HeapSnapshot_1.HeapSnapshotCapability,
    evalToString: Eval_1.EvalToStringCapability,
}, 'AppDebugCapabilities');
exports.DebugCapabilities = t.partial({
    app: exports.AppDebugCapabilities,
    companion: exports.AppDebugCapabilities,
    settings: exports.AppDebugCapabilities,
});
exports.ApplicationHostCapabilities = t.partial({
    install: AppInstall_1.AppInstallCapabilities,
    launch: Launch_1.LaunchCapabilities,
    screenshot: Screenshot_1.ScreenshotCapabilities,
    debug: exports.DebugCapabilities,
}, 'ApplicationHostCapabilities');
exports.HostCapabilities = t.partial({
    protocol: Meta_1.ProtocolCapabilities,
    io: BulkData_1.IOCapabilities,
    appHost: exports.ApplicationHostCapabilities,
    experimental: t.any,
}, 'HostCapabilities');
exports.InitializeResult = t.intersection([
    t.interface({
        device: t.string,
        hostKind: exports.HostKind,
        capabilities: exports.HostCapabilities,
    }),
    t.partial({
        hostID: t.string,
    }),
], 'InitializeResult');
