// apr-map/dist/apr-map.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('lodash.defaults'), require('apr-engine-each'), require('apr-engine-sum')) :
	typeof define === 'function' && define.amd ? define('apr-map', ['exports', 'lodash.defaults', 'apr-engine-each', 'apr-engine-sum'], factory) :
	(factory((global['apr-map'] = {}),global.defaults,global.aprEngineEach,global.Sum));
}(this, (function (exports,defaults,aprEngineEach,Sum) { 'use strict';

defaults = defaults && defaults.hasOwnProperty('default') ? defaults['default'] : defaults;
aprEngineEach = aprEngineEach && aprEngineEach.hasOwnProperty('default') ? aprEngineEach['default'] : aprEngineEach;
Sum = Sum && Sum.hasOwnProperty('default') ? Sum['default'] : Sum;

var run = function (ctx) {
  var sum = Sum(ctx.input);

  return aprEngineEach(defaults({
    after: function after(value, item) {
      sum[item.key] = value;
    }
  }, ctx)).then(function () {
    return sum;
  });
};

/**
 * @kind function
 * @name limit
 * @memberof map
 * @param {Array|Object|Iterable} input
 * @param {Number} limit
 * @param {Function} iteratee
 * @returns {Promise}
 */
var limit = function (input, limit, fn, opts) {
  return run({
    input: input,
    fn: fn,
    opts: defaults({
      limit: limit
    }, opts)
  });
};

/**
 * @kind function
 * @name series
 * @memberof map
 * @param {Array|Object|Iterable} input
 * @param {Function} iteratee
 * @returns {Promise}
 */
var series = function (input, fn, opts) {
  return limit(input, 1, fn, opts);
};

/**
 * <a id="map"></a>
 * Produces a new collection of values by mapping each value in `coll` through the `iteratee` function.
 *
 * [![](https://img.shields.io/npm/v/apr-map.svg?style=flat-square)](https://www.npmjs.com/package/apr-map) [![](https://img.shields.io/npm/l/apr-map.svg?style=flat-square)](https://www.npmjs.com/package/apr-map)
 *
 * @kind function
 * @name map
 * @param {Array|Object|Iterable} input
 * @param {Function} iteratee
 * @returns {Promise}
 *
 * @example
 * import awaitify from 'apr-awaitify';
 * import map from 'apr-map';
 *
 * const stat = awaitify(fs.stat);
 * const files = [
 *   'file1',
 *   'file2',
 *   'file3'
 * ];
 *
 * const stats = await map(files, async (file) =>
 *   await stat(file);
 * );
 */
var map = function (input, fn, opts) {
  return run({
    input: input,
    fn: fn,
    opts: opts
  });
};

var series$2 = series;
var limit$2 = limit;

map.series = series$2;
map.limit = limit$2;

exports['default'] = map;
exports.series = series$2;
exports.limit = limit$2;

Object.defineProperty(exports, '__esModule', { value: true });

})));
