// @fitbit/sdk-cli/lib/models/HostConnections.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fdb_debugger_1 = require("@fitbit/fdb-debugger");
const ts_events_1 = require("ts-events");
const developerRelay = tslib_1.__importStar(require("../api/developerRelay"));
class HostConnection {
    constructor(ws, host) {
        this.ws = ws;
        this.host = host;
    }
    static async connect(hostID) {
        const ws = await developerRelay.connect(hostID);
        return new this(ws, await fdb_debugger_1.RemoteHost.connect(ws));
    }
}
exports.HostConnection = HostConnection;
class HostConnections {
    constructor() {
        this.onHostAdded = new ts_events_1.SyncEvent();
    }
    async connect(hostType, hostID) {
        const existingHost = this[hostType];
        if (existingHost)
            existingHost.ws.destroy();
        const hostConnection = await HostConnection.connect(hostID);
        this[hostType] = hostConnection;
        this.onHostAdded.post({ hostType, host: hostConnection.host });
        return hostConnection;
    }
}
exports.default = HostConnections;
