// @fitbit/sdk/lib/collectComponentSourceMaps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const stream_1 = tslib_1.__importDefault(require("stream"));
const vinyl_1 = tslib_1.__importDefault(require("vinyl"));
const pathUtils_1 = require("./pathUtils");
function collectComponentSourceMaps() {
    const emitter = new stream_1.default.PassThrough({ objectMode: true });
    const collector = (componentType, componentPlatform) => new stream_1.default.Transform({
        objectMode: true,
        transform(file, _, next) {
            if (file.sourceMap) {
                const componentMapKey = [componentType];
                if (componentPlatform)
                    componentMapKey.push(componentPlatform);
                componentMapKey.push(pathUtils_1.normalizeToPOSIX(file.relative));
                const mapPath = ['sourceMaps', ...componentMapKey].join('/') + '.map';
                emitter.push(new vinyl_1.default({
                    componentMapKey,
                    contents: Buffer.from(JSON.stringify(file.sourceMap, undefined, 2), 'utf8'),
                    base: process.cwd(),
                    path: path_1.default.resolve(process.cwd(), mapPath),
                }));
            }
            next(undefined, file);
        },
    });
    return {
        collector,
        emitter,
    };
}
exports.default = collectComponentSourceMaps;
