// @fitbit/sdk/lib/componentTargets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ComponentType;
(function (ComponentType) {
    ComponentType["DEVICE"] = "device";
    ComponentType["COMPANION"] = "companion";
    ComponentType["SETTINGS"] = "settings";
})(ComponentType = exports.ComponentType || (exports.ComponentType = {}));
const componentTarget = {
    [ComponentType.DEVICE]: {
        inputs: ['app/index.ts', 'app/index.js'],
        output: 'app/index.js',
        notFoundIsFatal: true,
        translationsGlob: 'app/i18n/*.po',
    },
    [ComponentType.COMPANION]: {
        inputs: ['companion/index.ts', 'companion/index.js'],
        output: 'companion.js',
        notFoundIsFatal: false,
        translationsGlob: 'companion/i18n/*.po',
    },
    [ComponentType.SETTINGS]: {
        inputs: ['tsx', 'ts', 'jsx', 'js'].map((ext) => `settings/index.${ext}`),
        output: 'settings.js',
        notFoundIsFatal: false,
        translationsGlob: 'settings/i18n/*.po',
    },
};
exports.default = componentTarget;
