// @fitbit/sdk/lib/DiagnosticList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const diagnostics_1 = require("./diagnostics");
class DiagnosticList {
    constructor(diagnostics = []) {
        this.diagnostics = diagnostics;
        this.fatalError = false;
        if (!Array.isArray(diagnostics)) {
            throw new TypeError(`diagnostics must be array, not ${typeof diagnostics}`);
        }
    }
    extend(otherDiagnosticList) {
        this.diagnostics.push(...otherDiagnosticList.diagnostics);
        if (otherDiagnosticList.fatalError)
            this.fatalError = true;
    }
    push(diagnostic) {
        if (diagnostic.messageText === undefined ||
            diagnostic.category === undefined) {
            throw new TypeError('Diagnostic is missing `messageText` or `category` keys');
        }
        this.diagnostics.push(diagnostic);
    }
    pushMessage(text) {
        this.push({
            messageText: text,
            category: diagnostics_1.DiagnosticCategory.Message,
        });
    }
    pushWarning(text) {
        this.push({
            messageText: text,
            category: diagnostics_1.DiagnosticCategory.Warning,
        });
    }
    pushError(text) {
        this.push({
            messageText: text,
            category: diagnostics_1.DiagnosticCategory.Error,
        });
    }
    pushFatalError(text) {
        this.pushError(text);
        this.fatalError = true;
    }
}
exports.default = DiagnosticList;
