// apr-engine-iterator/dist/apr-engine-iterator.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('build-array'), require('lodash.isarraylike'), require('lodash.isfinite'), require('lodash.keys'), require('apr-engine-until')) :
	typeof define === 'function' && define.amd ? define('apr-engine-iterator', ['build-array', 'lodash.isarraylike', 'lodash.isfinite', 'lodash.keys', 'apr-engine-until'], factory) :
	(global['apr-engine-iterator'] = factory(global.buildArray,global.lodash,global.lodash$1,global.lodash$2,global.aprEngineUntil));
}(this, (function (buildArray,lodash,lodash$1,lodash$2,aprEngineUntil) { 'use strict';

buildArray = buildArray && buildArray.hasOwnProperty('default') ? buildArray['default'] : buildArray;
lodash = lodash && lodash.hasOwnProperty('default') ? lodash['default'] : lodash;
lodash$1 = lodash$1 && lodash$1.hasOwnProperty('default') ? lodash$1['default'] : lodash$1;
lodash$2 = lodash$2 && lodash$2.hasOwnProperty('default') ? lodash$2['default'] : lodash$2;
aprEngineUntil = aprEngineUntil && aprEngineUntil.hasOwnProperty('default') ? aprEngineUntil['default'] : aprEngineUntil;

var wrap = function wrap(fn) {
  var i = 0;

  var group = function group(size) {
    if (lodash$1(size)) {
      return buildArray(size).map(next);
    }

    return aprEngineUntil(function () {
      return next();
    });
  };

  var next = function next(size) {
    if (size) {
      return group(size);
    }

    return fn(i++);
  };

  return {
    next: next
  };
};

var arr = function arr(input) {
  return wrap(function (i) {
    return {
      value: input[i],
      done: input.length === i,
      key: i
    };
  });
};

var ittr = function ittr(input) {
  return wrap(function (i) {
    var item = input.next();

    return {
      done: item.done,
      value: item.value,
      key: i
    };
  });
};

var obj = function obj(input) {
  var okeys = lodash$2(input);

  return wrap(function (i) {
    var key = okeys[i];

    return {
      value: input[key],
      done: okeys.length === i,
      key: key
    };
  });
};

var engineIterator = function (input) {
  if (lodash(input)) {
    return arr(input);
  }

  if (input[Symbol.iterator]) {
    return ittr(input[Symbol.iterator]());
  }

  return obj(input);
};

return engineIterator;

})));
